apps/extension/src/routes/popup/settings/settings-connected-sites/index.tsx

```diff
@@ -0,0 +1,73 @@
+import { MagnifyingGlassIcon } from '@radix-ui/react-icons';
+import { Input } from '@penumbra-zone/ui/components/ui/input';
+import { LinkGradientIcon } from '../../../../icons/link-gradient';
+import { OriginRecord } from '@penumbra-zone/storage/src/chrome/types';
+import { AllSlices, useStore } from '../../../../state';
+import { UserChoice } from '@penumbra-zone/types/src/user-choice';
+import { SettingsScreen } from '../settings-screen';
+import { useStoreShallow } from '../../../../utils/use-store-shallow';
+import { allSitesFilteredOutSelector } from '../../../../state/connected-sites';
+import Map from '@penumbra-zone/polyfills/src/Map.groupBy';
+import { KnownSitesGroup } from './known-sites-group';
+
+const settingsConnectedSitesSelector = (state: AllSlices) => ({
+  filter: state.connectedSites.filter,
+  setFilter: state.connectedSites.setFilter,
+  knownSites: state.connectedSites.knownSites,
+});
+
+const getGroupedSites = (knownSites: OriginRecord[], filter?: string) => {
+  const groupedSites = Map.groupBy(knownSites, ({ choice }) => choice);
+  const filterFn = (site: OriginRecord) => !filter || site.origin.includes(filter);
+
+  return {
+    approvedSites: groupedSites.get(UserChoice.Approved)?.filter(filterFn) ?? [],
+    deniedSites: groupedSites.get(UserChoice.Denied)?.filter(filterFn) ?? [],
+    ignoredSites: groupedSites.get(UserChoice.Ignored)?.filter(filterFn) ?? [],
+  };
+};
```

Comment:
This is a helper function rather than a selector because, as a selector, it would force a re-render every time the state changes. That's because these arrays get redefined every time the state changes, so even `useStoreShallow` would detect changes to state and re-render.